src/utils/merge-nodes-with-matching-flavors.ts

```diff
@@ -0,0 +1,166 @@
+    mergeableImportFlavors,
+} from '../constants';
+import type { MergeNodesWithMatchingImportFlavors } from '../types';
+import { getImportFlavorOfNode } from './get-import-flavor-of-node';
+
+type MergeableFlavor = typeof mergeableImportFlavors[number];
+function isMergeableFlavor(flavor: string): flavor is MergeableFlavor {
+    return mergeableImportFlavors.includes(flavor as MergeableFlavor);
+}
+
+function selectMergeableNodesByImportFlavor(
+    nodes: ImportDeclaration[],
+): Record<MergeableFlavor, ImportDeclaration[]> {
+    return nodes.reduce(
+        (groups, node) => {
+            const flavor = getImportFlavorOfNode(node);
+            if (isMergeableFlavor(flavor)) {
+                groups[flavor].push(node);
+            }
+            return groups;
+        },
+        {
+            [importFlavorRegular]: [] as ImportDeclaration[],
+            [importFlavorType]: [] as ImportDeclaration[],
+        },
+    );
+}
+
+function selectNodeImportSource(node: ImportDeclaration) {
+    return node.source.value;
+}
+
+function nodeIsImportNamespaceSpecifier(
+    node: ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier,
+): node is ImportNamespaceSpecifier {
+    return node.type === 'ImportNamespaceSpecifier';
+}
+function nodeIsImportDefaultSpecifier(
+    node: ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier,
+): node is ImportDefaultSpecifier {
+    return node.type === 'ImportDefaultSpecifier';
+}
+
+/** Return false if the merge will produce an invalid result */
+function mergeIsSafe(
+    nodeToKeep: ImportDeclaration,
+    nodeToForget: ImportDeclaration,
+) {
+    if (
+        nodeToKeep.specifiers.some(nodeIsImportNamespaceSpecifier) ||
+        nodeToForget.specifiers.some(nodeIsImportNamespaceSpecifier)
+    ) {
+        // An `import * as Foo` namespace specifier cannot be merged
+        //   with other import expressions.
+        return false;
+    }
+    if (
+        nodeToKeep.specifiers.some(nodeIsImportDefaultSpecifier) &&
+        nodeToForget.specifiers.some(nodeIsImportDefaultSpecifier)
+    ) {
+        // Two `import Foo from` specifiers cannot be merged trivially.
+        // -- Notice: this is *not* import {default as Foo1, default as Foo2} -- that's legal!
+        //
+        // Future work could convert `import Foo1 from 'a'; import Foo2 from 'a';
+        //  into `import {default as Foo1, default as Foo2} from 'a';`
+        // But since this runs the risk of making code longer, this won't be in v1.
+        return false;
+    }
+    return true;
+}
+
+/**
+ * @returns (true to delete node | false to keep node)
+ */
+function mergeNodes(
+    nodeToKeep: ImportDeclaration,
+    nodeToForget: ImportDeclaration,
+) {
+    if (!mergeIsSafe(nodeToKeep, nodeToForget)) {
+        return false;
+    }
+
+    nodeToKeep.specifiers.push(...nodeToForget.specifiers);
+
+    // The line numbers will be all messed up. Is this a problem?
+    nodeToKeep.leadingComments = [
```

Comment:
I'm wondering what exactly people would expect how comments are merged. Consider

```js
// comment 1
// comment 2
import { foo } from "a"
// comment 3
// comment 4
// comment 5
// comment 6
import { bar } from "a"
// comment 7
// comment 8
```

which results in

```js
// comment 1
// comment 2
// comment 3
// comment 4
// comment 5
// comment 6
import { foo, bar  } from "a"
// comment 7
// comment 8
```

If one deletes the second import manually and adds the import to the first, one would get

```js
// comment 1
// comment 2
import { foo, bar } from "a"
// comment 3
// comment 4
// comment 5
// comment 6
// comment 7
// comment 8
```

I can see an argument for both cases. Though I'd assume comments in the import section are rare in general and if they do occur, they are probably meant as an explanation for a particular import, so keeping comments on top sound like the right thing to do.

Reply:
I think that the behavior shown as current (all comments go above) is the "more correct" option.  If that's what the code is doing now, I say we stick with it.

Reply:
I think the current behavior is reasonable, but maybe unintuitive. I think the root cause is that the parser prefers to assign comments to the `leading` position than the `trailing` position?

So comments 3, 4, 5, 6 are `leading` comments for the 2nd import, and my code moves them to be `leading` comments for the 1st import.

If comments 3-6 were considered `trailing` comments for the 1st import, then the behavior would have matched the `manual` process and maybe been a hair more intuitive?

Again, I agree the way it behaves now is probably the "more correct" option even if it doesn't match how a human would visualize the operations. Leading vs trailing comments is a weird thing to specify tbh.

Reply:
/shrug it's intuitive to me, if we're combining the comments above each of the imports to remain above the combined import.  I also think this is a super rare edge case, and probably not worth spending too much effort on perfecting.  ;)